polygeist/tools/cgeist/Lib/Attributes.h

```diff
@@ -9,90 +9,216 @@
+
+  /// Add \p Attr to the \p PassThroughAttr list.
+  /// Note: \p PassThroughAttr must have name "passthrough" and value of type
+  /// ArrayAttr.
+  static void addToPassThroughAttr(mlir::NamedAttribute &PassThroughAttr,
+                                   mlir::NamedAttribute Attr,
+                                   mlir::MLIRContext &Ctx);
+
+  /// Add \p NewAttrs into the attribute list of \p PassThroughAttr.
+  /// Note: \p PassThroughAttr must have name "passthrough" and value of type
+  /// ArrayAttr.
+  static void addToPassThroughAttr(mlir::NamedAttribute &PassThroughAttr,
+                                   mlir::ArrayAttr NewAttrs,
+                                   mlir::MLIRContext &Ctx);
+
+  /// Remove an attribute from the builder (if present).
+  /// Note: the given attribute will be removed even if it is contained by the
+  /// 'passthrough' named attribute.
+  AttrBuilder &removeAttribute(llvm::StringRef AttrName);
+  AttrBuilder &removeAttribute(llvm::Attribute::AttrKind Kind);
 
   /// Return true if the builder contains the specified attribute.
-  bool contains(llvm::StringRef attrName) const;
-  bool contains(llvm::Attribute::AttrKind kind) const;
-
-  /// Return true if the builder contains the specified attribute within the
-  /// 'passthrough' attribute.
-  bool containsInPassThrough(llvm::StringRef attrName) const;
-  bool containsInPassThrough(llvm::Attribute::AttrKind kind) const;
+  /// Note: these member functions also lookup for the given attribute in the
+  /// 'passthrough' named attribute if it exists.
+  bool contains(llvm::StringRef AttrName) const;
+  bool contains(llvm::Attribute::AttrKind Kind) const;
 
   /// Return true if the builder contains any attribute and false otherwise.
-  bool hasAttributes() const { return !attrs.empty(); }
+  bool hasAttributes() const { return !Attrs.empty(); }
 
   /// Return the given attribute if the builder contains it and llvm::None
   /// otherwise.
-  llvm::Optional<mlir::NamedAttribute> getAttr(llvm::StringRef attrName) const;
+  llvm::Optional<mlir::NamedAttribute> getAttr(llvm::StringRef AttrName) const;
   llvm::Optional<mlir::NamedAttribute>
-  getAttr(llvm::Attribute::AttrKind kind) const;
+  getAttr(llvm::Attribute::AttrKind Kind) const;
 
   /// Returns the attributes contained in the builder.
-  llvm::ArrayRef<mlir::NamedAttribute> getAttrs() const { return attrs; }
+  llvm::ArrayRef<mlir::NamedAttribute> getAttrs() const { return Attrs; }
+
+  mlir::MLIRContext &getContext() const { return Ctx; }
+
+  /// returns a NamedAttribute with name \p AttrName , and value \p Attr.
+  static mlir::NamedAttribute createNamedAttr(mlir::StringAttr AttrName,
+                                              mlir::Attribute Attr);
+
+  /// Returns a StringAttr of the form 'AttrName'.
+  static mlir::StringAttr createStringAttr(llvm::Twine AttrName,
+                                           mlir::MLIRContext &Ctx);
```

Comment:
Why use this over static `mlir::StringAttr` function?

Reply:
Removed.